### DISCUSSION TITLE
Discrepancy between compute times for prediction and clustering between 3.6.0 and 3.7.0 on databricks (pyspark)

### DISCUSSION THREAD

**Discussion by cvandenende** [moj-analytical-services/splink#1134] [General]
Hi all,

I noticed a discrepancy between the compute times of the prediction and clustering between the 3.7.0 and the 3.6.0 releases of Splink. I used the previous release of Splink (3.6.0) without problems on our databricks clusters, with approximately 90million comparisons the compute time for the prediction was +/- 25 minutes. Today, I ran the same code and data using the new splink release (3.7.0) and noticed a significant difference in the compute time for the same prediction. On 3.7.0 it takes +/- 1.5 hours and the cluster calculation could not complete (killed it after 2.5+ hours of computing). On 3.60 the clustering took around a minute, using the same settings, model and ingested data. When we specified version 3.6.0 the predictions and clustering run as expected. Has anyone else noticed something similar with the new release?

**Comment by RobinL** [comment]
@cvandenende Thanks for the report. We changed the way caching worked in version 3.7 but it sounds like this has broken caching for you in databricks.  We will investigate. For the moment continue to use 3.6.

If you're able to share any of your code where you're experiencing the problem it would be useful, but appreciate that it's not always possible

I have opened this as an issue tracking here:
https://github.com/moj-analytical-services/splink/issues/1136

**Reply by cvandenende** [reply]
Hi Robin, thanks for the quick reply. I can share the code, I have attached the databricks notebook we are using if you would like to have a look. Thanks in advance.
[TestSplink.py.txt](https://github.com/moj-analytical-services/splink/files/11037995/TestSplink.py.txt)

**Reply by RobinL** [reply]
Thanks very much.  An update.  There were two separate issues, 
- One to do with caching, fixed [here](https://github.com/moj-analytical-services/splink/pull/1137).  This has just been released in version `3.7.1`.  This fixes the poor clustering performance
- One to do with repartitioning, PR [here](https://github.com/moj-analytical-services/splink/pull/1138) which could explain the poorer `predict()` performance.  I am less sure about this, but in my tests, this new code is now faster than 3.6.0  Update: This is now released in version `3.7.2`.

On the repartitioning, there are two things which you could try to see what gives you best performance:

(1).  Try setting the `materialise_after_computing_term_frequencies=False` on `predict()` (see [here](https://github.com/moj-analytical-services/splink/blob/7ddb64c1bf20b6d8c593d60062cf69d6c84ab790/splink/linker.py#L1234).  That was the behaviour in 3.6.0, but the default has changed to `True` in 3.7.x.
(2).  Try different values of  `num_partitions_on_repartition ` (see [here](https://github.com/moj-analytical-services/splink/blob/7ddb64c1bf20b6d8c593d60062cf69d6c84ab790/splink/spark/spark_linker.py#L79)).  In version 3.7.2 (just released), I have changed it so that by default, it uses the value of either `spark.sql.shuffle.partitions` or `spark.default.parallelism`, but you can set it explicitly.  As a very rough guide, for 1m input records I'd go with `spark.sql.shuffle.partitions = 20`, and for 10m input records, maybe `spark.sql.shuffle.partitions = 100`.  But it depends on how tight your blocking rules are.

You change these settings like so:
```

conf = SparkConf()
conf.set("spark.sql.shuffle.partitions", "20")
conf.set("spark.default.parallelism", "20")

sc = SparkContext.getOrCreate(conf=conf)
spark = SparkSession(sc)
```


I'd be grateful for any feedback on whether these updates resolve your performance problems.

BTW - if you're from a public sector organisation (I believe you are), feel free to reach out to me.  It's great to hear you're using Splink.  I'd be happy to have a chat about a few ways you might be able to optimise your job for better runtimes.  The config looks good to me in terms of accuracy!  If you email me at 
![image](https://user-images.githubusercontent.com/2608005/226981389-307e0684-4b0b-482f-8871-fd82c4186dc4.png)
, I'll then respond from my work email.  If nothing else, it's nice for us to hear about how others are getting value from Splink.

**Reply by cvandenende** [reply]
Hi Robin, thanks again for all your work. This is very helpful and I will update with how it goes. I have also sent you an e-mail :)

**Reply by cvandenende** [reply]
Just a small update: Ran the same notebook with 3.7.2 and no other changes. The predict function still takes longer than expected (1.3hours) but the clustering seems to be fixed. Will test with the conf settings and report on how it does.

Edit: Adding materialise_after_computing_term_frequencies=False to the predict seems to have fixed the issue, our compute time with this setting is 16minutes (looks to be an improvement over 3.6.0).

**Reply by RobinL** [reply]
Excellent news, thanks for the update!  I'll reply to you're email soon 🙂

**Reply by RobinL** [reply]
A little more information on the options you can try in Databricks that may affect runtimes.

Note that any runtime tests must be done 'from cold'.   For example, 
- If you've run a training step, Splink will have already performed and cached some calculations, which will be re-used on `predict()`
- If you run `linker.predict()` twice, the first time it will put it in a cache, so the second time will run much faster because it will simply be retrieved from cache


Here are some of the settings you can tweak:


- Setting `materialise_after_computing_term_frequencies=False` will reduce the number of times Splink 'breaks lineage' (i.e. persists an intermediate result to disk to ensure it scales well).  This may be faster on smaller data, but breaking lineage is recommended for larger data
- When you instantiate the Spark linker, you can explicitly set `num_partitions_on_repartition` see [here](https://github.com/moj-analytical-services/splink/blob/41f9368753225f2fa560b09e12ceeec1839117bc/splink/spark/spark_linker.py#LL79C9-L79C38).  If left blank, it will guess a value from the value of of the `spark.sql.shuffle.partitions` or `spark.default.parallelism` Spark settings see [here](https://spark.apache.org/docs/latest/sql-performance-tuning.html).  You could try (say) 20, 100, and 400 to see what gives you best performance.  
- Setting `spark.sql.shuffle.partitions` to different values (e.g. 20, 100 and 400).  This is separate (does something different) to the `num_partitions_on_repartition` 
- Setting a `threshold_match_probability` on `predict()`, of say `threshold_match_probability=0.1`  (I realise in your testing this seems to have slowed things down - but it might be worth trying again - see below for a further note)
- You might want to try setting 'break_lineage_method' to `'parquet'`.  In databricks, the default is `'delta_lake_table'`  See [here](`https://github.com/moj-analytical-services/splink/blob/41f9368753225f2fa560b09e12ceeec1839117bc/splink/spark/spark_linker.py#L92`) and [here](https://github.com/moj-analytical-services/splink/blob/41f9368753225f2fa560b09e12ceeec1839117bc/splink/spark/spark_linker.py#LL246C42-L246C58)

If you're interested, there are various ways can now retrieve the SQL that was executed.  Some are documented [here](https://moj-analytical-services.github.io/splink/dev_guides/debug_modes.html).

Re: performance issues setting a threshold on `predict()`, I've opened a possible bug report [here](https://github.com/moj-analytical-services/splink/issues/1163)

No worries if you're happy with current performance, but let us know if you get around to trying any of the above and whether anything helps

Finally, I've looked at the changes between 3.6 and 3.7.   To replicate the same settings as 3.6 as closely as possible I believe you would need to set:
`materialise_after_computing_term_frequencies=False`
and also do not change `spark.default.parallelism` or `spark.sql.shuffle.partitions` and set `num_partitions_on_repartition` to 100

**Reply by cvandenende** [reply]
Hi @RobinL, thanks again for all the helpful input you've provided. I will update this post as we perform our testing.

**Reply by cvandenende** [reply]
Hi @RobinL we noticed another difference between 3.6.0 and the newer versions (everything after 3.7.0) on spark (databricks). In 3.6.0 the cluster.physical_name and predict.physical_name tables are written correctly to whatever db is specified. We can query those tables just fine, however, in 3.7+ splink does not materialise the cluster.physical_name (splink_df_representatives_hash) table. If we look at the db, splink appears to produce many splink_df_representatives_x_hash where x is any number followed by a hash code, however, the cluster.physical_name table is not materialised in the database we tested this on 3.7.2 and now on 3.8.1.

**Reply by RobinL** [reply]
Good spot - thanks -  I've raised a PR here which will be merged soon:
https://github.com/moj-analytical-services/splink/pull/1228
